miniapps/nurbs/nurbs_patch_ex1.cpp

```diff
@@ -0,0 +1,340 @@
+         // knot span.
+         for (int i=0; i<dim; ++i)
+         {
+            ir1D[i] = ir->ApplyToKnotIntervals(*kv[i]);
+         }
+
+         patchRule->SetPatchRules1D(p, ir1D);
+      }  // loop (p) over patches
+
+      patchRule->Finalize(mesh);
+      di->SetNURBSPatchIntRule(patchRule);
+   }
+
+   // 10. Assemble and solve the linear system
+   cout << "Assembling system patch-wise and solving" << endl;
+   AssembleAndSolve(b, di, ess_tdof_list, pa, static_cond, algebraic_ceed, x);
+
+   delete patchRule;
+
+   // 11. Save the refined mesh and the solution. This output can be viewed
+   //     later using GLVis: "glvis -m refined.mesh -g sol.gf".
+   ofstream mesh_ofs("refined.mesh");
+   mesh_ofs.precision(8);
+   mesh.Print(mesh_ofs);
+   ofstream sol_ofs("sol.gf");
+   sol_ofs.precision(8);
+   x.Save(sol_ofs);
+
+   // 12. Send the solution by socket to a GLVis server.
+   if (visualization)
+   {
+      char vishost[] = "localhost";
+      int  visport   = 19916;
+      socketstream sol_sock(vishost, visport);
+      sol_sock.precision(8);
+      sol_sock << "solution\n" << mesh << x << flush;
+   }
+
+   // 13. Optionally assemble element-wise and solve the linear system, to
+   //     compare timings and compute relative error.
+   if (compareToElementWise)
+   {
+      Vector x_pw, x_ew;
+      x.GetTrueDofs(x_pw);
+
+      cout << "Assembling system element-wise and solving" << endl;
+      DiffusionIntegrator *d = new DiffusionIntegrator(one);
+      // Element-wise partial assembly is not supported on NURBS meshes, so we
+      // pass pa = false here.
+      AssembleAndSolve(b, d, ess_tdof_list, false, static_cond, algebraic_ceed,
+                       x);
+
+      x.GetTrueDofs(x_ew);
+
+      const double solNorm = x_ew.Norml2();
+      x_ew -= x_pw;
+
+      cout << "Element-wise solution norm " << solNorm << endl;
+      cout << "Relative error of patch-wise solution "
+           << x_ew.Norml2() / solNorm << endl;
+   }
+
+   // 14. Free the used memory.
+   if (delete_fec)
+   {
+      delete fec;
+   }
+
+   return 0;
+}
+
+// This function deletes bfi when the BilinearForm goes out of scope.
+void AssembleAndSolve(LinearForm & b, BilinearFormIntegrator * bfi,
+                      Array<int> const& ess_tdof_list, const bool pa,
+                      const bool static_cond, const bool algebraic_ceed,
+                      GridFunction & x)
+{
+   FiniteElementSpace *fespace = b.FESpace();
+   BilinearForm a(fespace);
+   if (pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+
+   a.AddDomainIntegrator(bfi);  // Takes ownership of bfi
+
+   StopWatch sw;
+   sw.Start();
+
+   // Assemble the bilinear form and the corresponding linear system, applying
+   // any necessary transformations such as: eliminating boundary conditions,
+   // applying conforming constraints for non-conforming AMR, static
+   // condensation, etc.
+   if (static_cond) { a.EnableStaticCondensation(); }
```

Comment:
Also, we shouldn't have static condensation option, right?

Reply:
Removed.